exporters/secret/exporter.go

```diff
@@ -0,0 +1,156 @@
+package secret
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"github.com/giantswarm/k8sclient/v5/pkg/k8srestconfig"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	fieldSelector = "type=kubernetes.io/tls"
+)
+
+type Config struct {
+	Namespaces []string
+}
+
+type Exporter struct {
+	cert      *prometheus.Desc
+	ctx       context.Context
+	k8sClient *kubernetes.Clientset
+	logger    micrologger.Logger
+
+	namespaces []string
+}
+
+func DefaultConfig() Config {
+	return Config{
+		Namespaces: []string{},
+	}
+}
+
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	e.logger.Log("info", "start collecting metrics")
+
+	// If no namespace whitelist is provided then we check all namespaces
+	if len(e.namespaces) == 0 {
+		allNamespaces, err := e.k8sClient.CoreV1().Namespaces().List(e.ctx, metav1.ListOptions{})
+		if err != nil {
+			e.logger.Log("error", microerror.Mask(err))
+		}
+
+		for _, ns := range allNamespaces.Items {
+			// We just need the namespace's name
+			nsName := ns.Name
+			e.namespaces = append(e.namespaces, nsName)
+		}
+	}
+
+	listOpts := metav1.ListOptions{
+		FieldSelector: fieldSelector,
+	}
+
+	// Range over namespaces
+	for _, namespace := range e.namespaces {
+		// Get secrets in namespace
+		secrets, err := e.k8sClient.CoreV1().Secrets(namespace).List(e.ctx, listOpts)
+		if err != nil {
+			e.logger.Log("error", microerror.Mask(err))
+		}
+		// Range over secrets
+		for _, secret := range secrets.Items {
+			err := e.calculateExpiry(ch, namespace, secret)
+			if err != nil {
+				e.logger.Log("error", microerror.Mask(err))
+			}
+		}
+	}
+
+	e.logger.Log("info", "finished collecting metrics")
+}
+
+func (e *Exporter) calculateExpiry(ch chan<- prometheus.Metric, namespace string, secret v1.Secret) error {
+	secretName := secret.Name
+	certBytes := secret.Data["tls.crt"]
```

Comment:
I feel like this needs some kind of error handling here - cert-manager certs will always be stored in `tls.crt`, but what about TLS secrets created by other tools?

Reply:
you can use this to avoid nil errors, if the map does not have the requested key the `ok` will be false`
```
if certBytes, ok := secret.Data["tls.crt"]; ok {
 // this will happen only if there are data under ""tls.crt"
}
```